src/geometry/manifold/charted_space.lean

```diff
@@ -407,7 +407,114 @@ by simp [atlas, charted_space.atlas]
+of charted space `M` and `M'` (with model space `H × H'`) by taking the products of the charts. Now,
+on `H × H'`, there are two charted space structures with model space `H × H'` itself, the one coming
+from `charted_space_self`, and the one coming from the product of the two `charted_space_self` on
+each component. They are equal, but not defeq (because the product of `id` and `id` is not defeq to
+`id`), which is bad as we know. This expedient of renaiming `H × H'` solves this problem. -/
+def model_prod (H : Type*) (H' : Type*) := H × H'
+
+section
+local attribute [reducible] model_prod
+
+instance model_prod_inhabited {α β : Type*} [inhabited α] [inhabited β] :
+  inhabited (model_prod α β) :=
+⟨(default α, default β)⟩
+
+instance (H : Type*) [topological_space H] (H' : Type*) [topological_space H'] :
+  topological_space (model_prod H H') :=
+by apply_instance
+
+/- Next lemma shows up often when dealing with derivatives, register it as simp. -/
+@[simp, mfld_simps] lemma model_prod_range_prod_id
+  {H : Type*} {H' : Type*} {α : Type*} (f : H → α) :
+  range (λ (p : model_prod H H'), (f p.1, p.2)) = set.prod (range f) univ :=
+by rw prod_range_univ_eq
+
+end
+
+/-- The product of two charted spaces is naturally a charted space, with the canonical
+construction of the atlas of product maps. -/
+instance prod_charted_space (H : Type*) [topological_space H]
+  (M : Type*) [topological_space M] [charted_space H M]
+  (H' : Type*) [topological_space H']
+  (M' : Type*) [topological_space M'] [charted_space H' M'] :
+  charted_space (model_prod H H') (M × M') :=
+{ atlas            :=
+    {f : (local_homeomorph (M×M') (model_prod H H')) |
+      ∃ g ∈ charted_space.atlas H M, ∃ h ∈ (charted_space.atlas H' M'),
+        f = local_homeomorph.prod g h},
+  chart_at         := (λ x: (M × M'), (charted_space.chart_at H x.1).prod (charted_space.chart_at H' x.2)),
+  mem_chart_source :=
+  begin
+    intro x,
+    simp only [local_homeomorph.prod_to_local_equiv, set.mem_prod, mem_chart_source, and_self, local_equiv.prod_source],
+  end,
+  chart_mem_atlas  :=
+  begin
+    intro x,
+    use (charted_space.chart_at H x.1),
+    split,
+    { apply chart_mem_atlas _, },
+    { use (charted_space.chart_at H' x.2), simp only [chart_mem_atlas, eq_self_iff_true, and_self], }
+  end
+}
+
+section prod_charted_space
+
 variables [topological_space H] [topological_space M] [charted_space H M]
+[topological_space H'] [topological_space M'] [charted_space H' M'] {x : M×M'}
+
+@[simp] lemma chart_of_prod_eq_prod_of_charts_coe :
+  (chart_at (model_prod H H') x : M × M' → model_prod H H')
+  = (prod.map (chart_at H x.fst) (chart_at H' x.snd)) := rfl
+
+@[simp] lemma chart_of_prod_eq_prod_of_charts_coe_symm :
+  ((chart_at (model_prod H H') x).symm : model_prod H H' → M × M') =
+  (prod.map (chart_at H x.fst).symm (chart_at H' x.snd).symm) := rfl
+
+@[simp] lemma chart_of_prod_eq_prod_of_charts_coe_to_local_equiv_trans {α : Type*} {β : Type*}
```

Comment:
You haven't fully felt the power of simp yet. You see how your proof is not straightforward, and you need to work and tell Lean everything. It is just because you are missing the simp lemma telling Lean what that the definition of `chart_at` in a product space is. Something like
```lean
@[simp, mfld_simps] lemma prod_charted_space_chart_at :
  (chart_at (model_prod H H') x) = (chart_at H x.fst).prod (chart_at H' x.snd) := rfl
```
Once you have this, the proof of this lemma just becomes `by simp` ! (and it probably means that the lemma is not necessary any more).

And I would also remove the two previous lemmas (with `prod.map`), as they wouldn't be useful any more either! Note that `prod.map` is not so nice currently to use in mathlib as it has no registered simp lemma, so I would avoid using it altogether (or add the supporting lemmas).

Reply:
What do you mean exactly by adding the supporting lemmas? I believe giving names is a very powerful thing in maths and in this case, it means that if someone sees `prod.map`, he/she recognizes it immediately and can immediately guess lemmas' names containing `prod_map` such the times_cont_diff one, whereas if these names are not used it is way harder to find the right lemmas one needs. For example, it took me some time to find the ones I needed before I knew that the term `prod.map` existed, and I had to ask on Zulip for it.

Reply:
If you apply `prod.map f g` to a point `p`, you get `(f p.1, g p.2)` by definition, but `simp` can not prove it. This means this simp lemma is missing. Without it, `prod.map` is not nearly as usable as it should be. So it is definitely a good idea to add it to mathlib, probably just after the definition of `prod.map`.

Reply:
Should I open a different PR to do it? It has nothing to do with product manifolds!

Reply:
Moreover removing the lemmas you say and adding yours does indeed work, and `simp` is able to do things without. However, when removing them in the proofs that standard product maps are smooth, squeeze_simp produces 5 lines instead of 3 and is slower! Do you think that this implies it is a good idea to leave them?

Reply:
It's just one lemma, so it would be fine in this PR. Often, in PRs on a specific topic there are a few related supporting lemmas scattered elsewhere around the library. If there are too many of them (say more than 10), then it is worth making a preliminary PR, otherwise it's fine.

Reply:
The fact that simp only produces a very long output is one of the reason I have introduced the simpset `mfld_simps`. If you start from `squeeze_simp with mfld_simps`, you should get a more reasonable output. Typically, `simp only with mfld_simps` will even work out of the box.

Reply:
> It's just one lemma, so it would be fine in this PR. Often, in PRs on a specific topic there are a few related supporting lemmas scattered elsewhere around the library. If there are too many of them (say more than 10), then it is worth making a preliminary PR, otherwise it's fine.

Ok, just to be sure, I will add this after the definition of `prod.map`:
```
@[simp] lemma {u₁ u₂ v₁ v₂} prod_map {α₁ : Type u₁} {α₂ : Type u₂} {β₁ : Type v₁} {β₂ : Type v₂}
  (f : α₁ → α₂) (g : β₁ → β₂) : prod.map f g = λ x, (f x.1, g x.2) := rfl
```

Reply:
The definition of `prod.map` is in `lean/library/data/prod` so I am forced to do a different PR as it is a different project. Correct me if I am wrong

Reply:
You can't add it right there because it is in core, not in mathlib. But you can add it at the end of `logic/function/basic`, for instance. And you can use instead
```
@[simp] lemma prod_map {α₁ α₂ β₁ β₂ : Type*}
  (f : α₁ → α₂) (g : β₁ → β₂) : prod.map f g = λ x, (f x.1, g x.2) := rfl
```
This is equivalent to your formulation, but doesn't specify explicitly the universes.

Reply:
I think I am not a contributor of `lean` as git does not let me branch it

Reply:
> You can't add it right there because it is in core, not in mathlib. But you can add it at the end of `logic/function/basic`, for instance. And you can use instead
> 
> ```
> @[simp] lemma prod_map {α₁ α₂ β₁ β₂ : Type*}
>   (f : α₁ → α₂) (g : β₁ → β₂) : prod.map f g = λ x, (f x.1, g x.2) := rfl
> ```
> 
> This is equivalent to your formulation, but doesn't specify explicitly the universes.

Ok great thanks!

Reply:
Why not doing an analogous lemma for `function.comp`?

Reply:
Because it already works:
```
example (α β γ : Type*) (f : α → β) (g : β → γ) (x : α) :
  function.comp g f x = g (f x) := by simp
```
The reason is that `function.comp` is marked as reducible, so `simp` will already transform `function.comp g f` into its definition.

Reply:
Ok and why not marking `prod.map` with reducible as well?

Reply:
I think I would, but since it's in core it harder to do.